### PR TITLE
[Coinmate] Adding Coinmate private endpoint websocket support

### PR DESCRIPTION
**Pull request by makarid**

### CODE REVIEW

**Comment by badgerwithagun** [comment]
Please note that I am unable to merge PRs at present due to an issue with Travis not notifying GitHub that the build ran successfully. The repository is set to disallow merges if checks are in progress/failing so merging is disabled.

This needs @dozd's attention.

**Comment by makarid** [comment]
Thanks a lot for your review @badgerwithagun . I will check all of these when i have the time.

File: service-pusher/src/main/java/info/bitrich/xchangestream/service/pusher/PusherStreamingService.java [hunk 1]
```diff
@@ -31,6 +32,13 @@ public PusherStreamingService(String apiKey, String cluster) {
         pusher = new Pusher(apiKey, options);
     }
 
+    public PusherStreamingService(String apiKey, String cluster, Authorizer authorizer) {
```
[the remark this answers is not in the document]
Comment (makarid) [reply service-pusher/src/main/java/info/bitrich/xchangestream/service/pusher/PusherStreamingService.java]: Fixed

File: xchange-coinmate/src/main/java/info/bitrich/xchange/coinmate/CoinmateStreamingAccountService.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+    private final PusherStreamingService service;
+    private final String userId;
+
+    public CoinmateStreamingAccountService(PusherStreamingService service, String userId) {
+        this.service = service;
+        this.userId = userId;
+    }
+
+    public Observable<Balance> getBalanceChanges(Currency currency, Object... args) {
```
[the remark this answers is not in the document]
Comment (makarid) [reply xchange-coinmate/src/main/java/info/bitrich/xchange/coinmate/CoinmateStreamingAccountService.java]: I believe i have fixed also

File: xchange-coinmate/src/main/java/info/bitrich/xchange/coinmate/CoinmateStreamingAdapter.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+    public static String getChannelPostfix(CurrencyPair currencyPair) {
+        return currencyPair.base.toString().toUpperCase() + "_" + currencyPair.counter.toString().toUpperCase();
+    }
+
+    public static UserTrades adaptWebSocketUserTrades(List<CoinmateWebSocketUserTrade> coinmateWebSocketUserTrades, CurrencyPair currencyPair) {
+        List<UserTrade> userTrades = new ArrayList<>();
+        coinmateWebSocketUserTrades.forEach((coinmateWebSocketUserTrade) -> {
+            userTrades.add(
+                    new UserTrade(
```
[the remark this answers is not in the document]
Comment (makarid) [reply xchange-coinmate/src/main/java/info/bitrich/xchange/coinmate/CoinmateStreamingAdapter.java]: Fixed

File: xchange-coinmate/src/main/java/info/bitrich/xchange/coinmate/CoinmateStreamingAdapter.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+            );
+        });
+        return new UserTrades(userTrades, Trades.TradeSortType.SortByTimestamp);
+    }
+
+    public static OpenOrders adaptWebsocketOpenOrders(List<CoinmateWebsocketOpenOrder> coinmateWebsocketOpenOrders, CurrencyPair currencyPair) {
+        List<LimitOrder> openOrders = new ArrayList<>();
+        coinmateWebsocketOpenOrders.forEach((coinmateWebsocketOpenOrder) -> {
+            openOrders.add(
+                    new LimitOrder(
```
[the remark this answers is not in the document]
Comment (makarid) [reply xchange-coinmate/src/main/java/info/bitrich/xchange/coinmate/CoinmateStreamingAdapter.java]: Fixed

File: xchange-coinmate/src/main/java/info/bitrich/xchange/coinmate/CoinmateStreamingTradeService.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    public Observable<Order> getOrderChanges(CurrencyPair currencyPair, Object... args) {
+        String channelName = "private-open_orders-" + userId + "-" + CoinmateStreamingAdapter.getChannelPostfix(currencyPair);
+
+        return service.subscribePrivateChannel(channelName, "open_orders")
+                .map((message) -> {
+                    List<CoinmateWebsocketOpenOrder> websocketOpenOrders =
+                            StreamingObjectMapperHelper.getObjectMapper().readValue(message, new TypeReference<List<CoinmateWebsocketOpenOrder>>() {});
+                    return CoinmateStreamingAdapter.adaptWebsocketOpenOrders(websocketOpenOrders, currencyPair);
+                })
+                .concatMapIterable(OpenOrders::getAllOpenOrders);
```
[the remark this answers is not in the document]
Comment (makarid) [reply xchange-coinmate/src/main/java/info/bitrich/xchange/coinmate/CoinmateStreamingTradeService.java]: I will fix this some other time

File: xchange-coinmate/src/main/java/info/bitrich/xchange/coinmate/CoinmateStreamingTradeService.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    public Observable<UserTrade> getUserTrades(CurrencyPair currencyPair, Object... args) {
+        String channelName = "private-user-trades-" + userId + "-" + CoinmateStreamingAdapter.getChannelPostfix(currencyPair);
+
+        return service.subscribePrivateChannel(channelName, "user_trades")
+                .map((message) -> {
+                    List<CoinmateWebSocketUserTrade> webSocketUserTrades =
+                            StreamingObjectMapperHelper.getObjectMapper().readValue(message, new TypeReference<List<CoinmateWebSocketUserTrade>>() {});
+                    return CoinmateStreamingAdapter.adaptWebSocketUserTrades(webSocketUserTrades, currencyPair);
+                })
+                .concatMapIterable(UserTrades::getUserTrades);
```
[the remark this answers is not in the document]
Comment (makarid) [reply xchange-coinmate/src/main/java/info/bitrich/xchange/coinmate/CoinmateStreamingTradeService.java]: I will fix this some other time, it it is ok

**Comment by makarid** [comment]
@badgerwithagun when you have time please check my updates. Thank you!

**Comment by makarid** [comment]
Thank you @badgerwithagun for your replies,edits and reviews.

**Pull request closed without merging**